core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/StandaloneDockerContainerFactory.scala

```diff
@@ -0,0 +1,100 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.openwhisk.core.containerpool.docker
+
+import akka.actor.ActorSystem
+import org.apache.commons.lang3.SystemUtils
+import org.apache.openwhisk.common.{Logging, TransactionId}
+import org.apache.openwhisk.core.WhiskConfig
+import org.apache.openwhisk.core.containerpool.{Container, ContainerFactory, ContainerFactoryProvider}
+import org.apache.openwhisk.core.entity.{ByteSize, ExecManifest, InvokerInstanceId}
+import pureconfig.loadConfig
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+
+object StandaloneDockerContainerFactoryProvider extends ContainerFactoryProvider {
+  override def instance(actorSystem: ActorSystem,
+                        logging: Logging,
+                        config: WhiskConfig,
+                        instanceId: InvokerInstanceId,
+                        parameters: Map[String, Set[String]]): ContainerFactory = {
+    val client =
+      if (SystemUtils.IS_OS_MAC) new DockerForMacClient()(actorSystem.dispatcher)(logging, actorSystem)
+      else if (SystemUtils.IS_OS_WINDOWS) new DockerForWindowsClient()(actorSystem.dispatcher)(logging, actorSystem)
+      else new DockerClientWithFileAccess()(actorSystem.dispatcher)(logging, actorSystem)
+
+    new StandaloneDockerContainerFactory(instanceId, parameters)(
+      actorSystem,
+      actorSystem.dispatcher,
+      logging,
+      client,
+      new RuncClient()(actorSystem.dispatcher)(logging, actorSystem))
+  }
+}
+
+class StandaloneDockerContainerFactory(instance: InvokerInstanceId, parameters: Map[String, Set[String]])(
+  implicit actorSystem: ActorSystem,
+  ec: ExecutionContext,
+  logging: Logging,
+  docker: DockerApiWithFileAccess,
+  runc: RuncApi)
+    extends DockerContainerFactory(instance, parameters) {
+  private val pulledImages = new TrieMap[String, Boolean]()
+
+  override def createContainer(tid: TransactionId,
+                               name: String,
+                               actionImage: ExecManifest.ImageName,
+                               userProvidedImage: Boolean,
+                               memory: ByteSize,
+                               cpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+
+    //For standalone server usage we would also want to pull the OpenWhisk provided image so as to ensure if
+    //local setup does not have the image then it pulls it down
+    //For standard usage its expected that standard images have already been pulled in.
+    val imageName = actionImage.localImageName(runtimesRegistryConfig.url)
+    val pulled = if (!userProvidedImage && !pulledImages.contains(imageName)) {
+      docker.pull(imageName)(tid).map { _ =>
```

Comment:
this will be problematic if one wants to use a build images that's not on dockerhub yet. 
a docker run will pull a missing image if necessary, maybe we can support the `bypass` options available on the invoker role today to skip pulling images. wdyt?